### PR TITLE
New branch naming convention

### PR DESCRIPTION
**Pull request by rigelrozanski**
Also totally re-worked the github section for increased clarity

### CODE REVIEW

**Comment by adrianbrink** [comment]
Let's include this article so that people can have a reference point: https://nvie.com/posts/a-successful-git-branching-model/

**Review by adrianbrink — APPROVED** [review]

File: go/coding_standard.md
Comment (jbibla) [inline go/coding_standard.md]: `structure`

File: go/coding_standard.md
Comment (jbibla) [inline go/coding_standard.md]: typo: `follow`
interested in adding ticket number? `jordan/666-my-branch-description`?

File: go/coding_standard.md
Comment (jbibla) [inline go/coding_standard.md]: typos:
"All feature PRs should be merged to the `develop` branch, unless stacked on other PRs leading to `develop`."

File: go/coding_standard.md
Comment (jbibla) [inline go/coding_standard.md]: typo: `it is` not `is is`

File: go/coding_standard.md
Comment (jbibla) [inline go/coding_standard.md]: i'd add a note about codecov always going up or staying neutral. we need to stop merging PRs when codecov is going down.

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: * Production repos never push to `master` or `develop` branches directly. Push-master can merge to these branches.

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: * `develop` should always pass CI and CD.

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: * should following the naming convention `feature/<branch-description>`

Assign it to the owner, instead of adding the owner name to the branch name.

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: I don't see the point of this. Cluttering the list of PRs with "WIP" PRs that stay there for weeks seems unprofessional. What is the benefit of a "WIP" PR?

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: I don't get this. Assuming there's a chain of PRs: `develop` <- `PR1` <- `PR2`. What happens with `PR2` when `PR1` is merged into `develop`? Is it going to be open against `develop`? If that's the case, this is an interesting way of stacking PRs.

Overall this seems like a workaround for the slowness of PR approvals. Instead of working around it, let's try to see how we can improve that and fix it.

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: Please do not use */ref/* in any case! It is an internally used keyword by GitHub and you can cause all kinds of havoc throwing it around. I think the correct naming convention would be `reference/<branch-description>` if something like this is necessary.
Assign it to the owner, instead of adding the owner name to the branch name.
Also, try to avoid multiple **/** in the naming conventions. Same thing, a lazily formatted automation will choke on it.

File: go/coding_standard.md
Comment (greg-szabo) [inline go/coding_standard.md]: Just for clarity: "If the PR is opened against the develop branch, then include the changes under an `unreleased changes` header at the top of the `CHANGELOG.md` document."

**Review by greg-szabo — CHANGES_REQUESTED** [review]
Good ideas. Improvement ideas were added in the comments.

File: go/coding_standard.md
Comment (melekes) [reply go/coding_standard.md]: I am using `666-my-branch-description`, so anyone can find branch I am working on by looking out issue number, typing 666 and pressing `TAB` (autofilling)

File: go/coding_standard.md
Comment (faboweb) [reply go/coding_standard.md]: This shows contributors that there is active work done on a certain ticket.

> PRs that stay there for weeks 

That those stay for weeks is the problem. A PR should be closed in 2-3 days. If not the scope might not be defined well.

File: go/coding_standard.md
Comment (faboweb) [reply go/coding_standard.md]: I think PRs should be independent and scopes should be defined so that this is possible.

File: go/coding_standard.md
Comment (faboweb) [reply go/coding_standard.md]: The released changes will be in the `CHANGELOG.md` on master right? I think we should add to `CHANGELOG.md` as if this is master (so every PR could be part of the release without trouble). I think having a `unreleased changes` section will cause more work/trouble.

**Comment by faboweb** [comment]
Solid set of rules 👍

File: go/coding_standard.md
Comment (jbibla) [reply go/coding_standard.md]: @greg-szabo i agree with you on this one, but many of the devs find it helpful...

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: An optional idea: adding the issue number in the feature description makes it easier to track issue/PR pairs. You might be able to do it by referencing the PR in the issue or vice versa, so I'm not sure if it's useful or not. Adding an "ID" to the description (Anton = 666) seems weird for me but not disturbing, so I don't have anything against it.

File: go/coding_standard.md
Comment (zramsay) [inline go/coding_standard.md]: but sometimes it's worth not squashing because the PR has a series of commits that ought to be kept for future reference

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: It seems all PMs agree :) (Fabo's comment that a PR should be closed within 2-3 days tell me it's really not a WIP, but a regular PR.)

File: go/coding_standard.md
Comment (zramsay) [reply go/coding_standard.md]: maybe works for UI code, but codecov is non-deterministic. For example, you can edit the README or docs in a repo, make a PR, and codecov will change. point being, this shouldn't be a requirement. docs aside, I've even seen a PR that was only tests being added that had codecov go down

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: Agreed, should be case-by-case.

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: This seems overly complicated and will lead to missing additions. Can't we compile the changelog additions from the delta since last merge to master? As we expect only completed PRs to be merged in and not single commits, this should make for a clear overview of what has changed. Down to that it's possible to do it in an automatic fashion.

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: Another usability problem is that CodeCov doesn't update the report in the PR frequently enough, if at all, after the initial run. Maybe something we can confiugre or improve in another way.

File: go/coding_standard.md
Comment (faboweb) [reply go/coding_standard.md]: What would you base the delta on? PRs often hold more then one change. And parsing all commit messages sounds like information overload. But I like the idea of automation. Maybe the PR has to state exactly what was done in a formatted way. This way it is also easier to read for approvers.

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: If done advocated like this, it will lead to long-running feature branches, which are prone to diverge to the point that siginificant energy needs to be spent to prepare them for merge. Instead of stacking the work should be split up in reasonable chunks that can build of one another but make for mergable changes to develop. High-level progress spanning PRs can always be tracked in a dedicated issue.

File: go/coding_standard.md
Comment (faboweb) [reply go/coding_standard.md]: I need to clarify: I like the "WIP" approach as it creates transparency about what is worked on, especially if the work takes longer then one day. The "WIP" flag then shows me, that I don't need to dive in and approve it yet.
But I think work on one item should in general be done after 3 days.

File: go/coding_standard.md
Comment (xla) [inline go/coding_standard.md]: What's the benefit here? It assumes there is a clear idea of a the next release a priori.

File: go/coding_standard.md
Comment (faboweb) [reply go/coding_standard.md]: Codecov will only consider actual code if configured correctly. Changing the README will result in no change in codecov and is therefor a valid PR.
I can only guess what happened in your example with the tests. I assume adding the tests imported code that was not imported before, therefor raising the overall code. If this new cide then is not 100% covered, codecov goes down.
I am happy to give it a look if this happens somewhere.

Comment (faboweb) [reply go/coding_standard.md]: @xla CodeCov should update after each run of the CI if configured correctly.

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: Indeed, I assume that both the title of the PR as well as the description communicate clearly what to expect from the change. The changelog should be a high-level index for the curious to get an overview and pointers where to look for more info, an example line:

* Concise description ([Link to PR](http://very-nice-pr.org))

Check the Hashicorp projects for inspiration: [Terraform Changelog](https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md).

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: @faboweb This hasn't been my experience, see tendermint/tendermint#1237 as an example where the description from Codegov is not up-to-date with the latest commits.

Comment (xla) [reply go/coding_standard.md]: We could also make it part of the acceptance stage of a PR much as our normal CI. [Commit Status](https://docs.codecov.io/docs/commit-status) should make that possible.

Comment (faboweb) [reply go/coding_standard.md]: All this is the case in the voyager repo.
@greg-szabo I assume you set this up. Do you want to take a shot at this? We can pair, if you need assistance.

Comment (xla) [reply go/coding_standard.md]: @faboweb Could you link to the repo you mentioned?

Comment (faboweb) [reply go/coding_standard.md]: Sorry here: https://github.com/cosmos/voyager

Comment (xla) [reply go/coding_standard.md]: Inspired and shamelessly ripped off in tendermint/tendermint#1247.

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: Do you think it's a bad idea to track high level comments here? Care to expand a bit on your thinking?

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: @greg-szabo correct

@faboweb I don't like the idea of imposing a structure as to how the comments must be created in the PR ........ If anything we could compile the description of commit messages to the changelog - but then you lose all the context of how the commit messages relate to each other. Not to mention ideas form dynamically and this would lead to manually having to edit old commit descriptions - less than idea. I think manual tracking is the only sensible solution

@xla nice! I like the Terraform changelog - I'm adamant that it is a good idea to track unreleased changes as done there but I still think they should be tracked manually as expressed in above comment ^

File: go/coding_standard.md
Comment (xla) [reply go/coding_standard.md]: My point is that the on-going PR doesn't add much value, create one at the time of the release and take the contents from the PRs merged into `develop` as basis for the release, no need to have two places to track, kept up-to-date.

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: thanks

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: darn that's what happens when I edit right in github and not in vim

Comment (rigelrozanski) [reply go/coding_standard.md]: addressed

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: addressed

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: addressed

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: addressed

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: Cool I'm going to leave this point for somebody else to explain opening an issue: https://github.com/tendermint/coding/issues/55

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: added

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: addressed (Continuous Integration, Continuous Delivery)

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: I like branches with owner names - easier to identify right off the bat and know who is responsible for what just by looking at the branch names

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: It's not always the case that all features can be created in such a short period of time, however I do agree that we should keep as small parcelized as possible. I think it's absolutely essential to have WIP in the title to keep track of what is in on-going development - this has already been established for a while now... if you're working on a branch, open a PR, add WIP

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: side note, assigned branches are totally unintuitive as far as I can tell - leaving as is, still open to being convinced but I don't think there is nice solution which doesn't require addons - maybe open a new PR to coding which reverts author names in the branch names and provides a nice way to view the list of all the branch owners for a repo that doesn't involve guessing games based on commits

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: addressed

File: go/coding_standard.md
Comment (rigelrozanski) [reply go/coding_standard.md]: Cool - reworking to the likes of your idea

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: I still don't like it, but in a "disagree and commit" fashion, let's do it. I don't see any harm, except to make sure to tell new hires that this is how we do it.

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: I still don't see the point of opening a PR to a branch that's not "ready" for review. (I see the point when the review says that there needs to be extra work done.) Instead of fighting semantics, I'm going to accept that this is how we do it. I hope it has real benefits for the dev team.

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: Ok, this is one that seems to require more discussions. @rigelrozanski do you have an example where this was useful? Did it create more confusion? Did it help in chunking your work?

My opinion is what @xla said above: independent PRs, that can build on one another but make for mergeable changes to develop.

To be honest, I would even add that commit cleanup is required before merging a PR:
1. All checkpoint commits should be squashed. (Commits that are created to "save" your work but would result in an unstable/unbuildable code.)
2. All regular ("stable") commits should go into the merge.

I'm not trying to enforce this, but it's a good coding habit.

I'm open for more information.

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: I can accept the `yourname/<whatev>` naming convention as I mentioned above.
As for the reference stuff, would it work to just add the keyword reference in the branch description? Something like `yourname/123-reference-coolfeatureidea` or even `yourname/reference-123-coolfeatureidea` or something without numbers if issues are not open for it. (Which hopefully doesn't happen. :) )

File: go/coding_standard.md
Comment (greg-szabo) [reply go/coding_standard.md]: I had nothing to do with CodeDev. :) Accepting that it's non-deterministic freaks me out a bit. Maybe for now it should just be optional...
I have more faith in the upcoming Jenkins testing, but that takes some time...

**Review by greg-szabo — APPROVED** [review]
Two things still to discuss:
1. PR stacking
2. reference branch naming convention

File: go/coding_standard.md
Comment (ebuchman) [reply go/coding_standard.md]: > Check the Hashicorp projects for inspiration: Terraform Changelog.

I think I copied their changelog style verbatim.

Automating this sounds hard, we'd have to be super disciplined in the PR descriptions. Maybe it's a long term ideal. In the meantime I suppose it should be ok to just get it from the set of merged PRs since the last release. Ideally folks would update the changelog [unreleased] as they go to simplify this but I'm worried about the merge conflicts

**Pull request merged**
